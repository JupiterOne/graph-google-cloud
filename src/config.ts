import {
  IntegrationExecutionContext,
  IntegrationLogger,
} from '@jupiterone/integration-sdk-core';
import { ServiceUsageClient } from './steps/service-usage/client';
import { IntegrationConfig, SerializedIntegrationConfig } from './types';
import { deserializeIntegrationConfig } from './utils/integrationConfig';
import { google } from 'googleapis';
import { handleApiClientError } from './google-cloud/client';
import { ServiceUsageListFilter } from './google-cloud/types';

export async function executeTestRequest(
  config: IntegrationConfig,
  logger: IntegrationLogger,
) {
  try {
    const client = google.serviceusage({ version: 'v1', retry: false });
    const googleClient = new ServiceUsageClient({ config }, logger);
    const auth = await googleClient.getAuthenticatedServiceClient();

    await client.services.list({
      parent: `projects/${config.serviceAccountKeyConfig.project_id}`,
      pageSize: 200,
      auth: auth,
      filter: ServiceUsageListFilter.ENABLED,
    });
  } catch (err) {
    throw handleApiClientError(err);
  }
}

export async function validateInvocation(
  context: IntegrationExecutionContext<SerializedIntegrationConfig>,
) {
  const { instance, logger } = context;
  const { config: serializedIntegrationConfig } = instance;
  const config = (context.instance.config = deserializeIntegrationConfig(
    serializedIntegrationConfig,
  ));

  await executeTestRequest(config, logger);
}
