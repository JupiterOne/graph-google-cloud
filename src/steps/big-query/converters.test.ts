import {
  createBigQueryDatasetEntity,
  createBigQueryModelEntity,
  createBigQueryTableEntity,
} from './converters';
import {
  getMockBigQueryDataset,
  getMockBigQueryModel,
  getMockBigQueryTable,
} from '../../../test/mocks';
import { DEFAULT_INTEGRATION_CONFIG_PROJECT_ID } from '../../../test/config';

describe('#createBigQueryDatasetEntity', () => {
  test('should convert to entity', () => {
    expect(
      createBigQueryDatasetEntity(getMockBigQueryDataset()),
    ).toMatchSnapshot();
  });

  test('should convert to entity with public param set to on', () => {
    expect(
      createBigQueryDatasetEntity(
        getMockBigQueryDataset({
          access: [
            { role: 'WRITER', specialGroup: 'allAuthenticatedUsers' },
            { role: 'WRITER', specialGroup: 'projectWriters' },
            { role: 'OWNER', specialGroup: 'projectOwners' },
            { role: 'OWNER', userByEmail: 'admin@creativice.com' },
            { role: 'READER', iamMember: 'allUsers' },
            { role: 'READER', specialGroup: 'projectReaders' },
          ],
        }),
      ),
    ).toMatchSnapshot();
  });
});

describe('#createBigQueryModelEntity', () => {
  test('should convert to entity', () => {
    expect(createBigQueryModelEntity(getMockBigQueryModel())).toMatchSnapshot();
  });
});

describe('#createBigQueryTableEntity', () => {
  test('should convert to entity', () => {
    expect(
      createBigQueryTableEntity({
        data: getMockBigQueryTable(),
        projectId: DEFAULT_INTEGRATION_CONFIG_PROJECT_ID,
        isPublic: false,
        kmsKeyName: 'example-kms-key-name',
      }),
    ).toMatchSnapshot();
  });
});
