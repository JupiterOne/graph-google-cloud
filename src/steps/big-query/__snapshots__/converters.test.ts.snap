// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createBigQueryDatasetEntity should convert to entity 1`] = `
{
  "_class": [
    "DataStore",
    "Database",
  ],
  "_key": "j1-gc-integration-dev-v2:test_big_query_dataset",
  "_rawData": [
    {
      "name": "default",
      "rawData": {
        "access": [
          {
            "role": "WRITER",
            "specialGroup": "projectWriters",
          },
          {
            "role": "OWNER",
            "specialGroup": "projectOwners",
          },
          {
            "role": "OWNER",
            "userByEmail": "admin@creativice.com",
          },
          {
            "role": "READER",
            "specialGroup": "projectReaders",
          },
        ],
        "creationTime": "1610919167542",
        "datasetReference": {
          "datasetId": "test_big_query_dataset",
          "projectId": "j1-gc-integration-dev-v2",
        },
        "etag": "Q1RJczpo1aoPRRoUJ1pElQ==",
        "id": "j1-gc-integration-dev-v2:test_big_query_dataset",
        "kind": "bigquery#dataset",
        "lastModifiedTime": "1610919344642",
        "location": "US",
        "selfLink": "https://www.googleapis.com/bigquery/v2/projects/j1-gc-integration-dev-v2/datasets/test_big_query_dataset",
      },
    },
  ],
  "_type": "google_bigquery_dataset",
  "classification": null,
  "createdOn": 1610919167542,
  "description": undefined,
  "displayName": "test_big_query_dataset",
  "encrypted": true,
  "etag": "Q1RJczpo1aoPRRoUJ1pElQ==",
  "id": "j1-gc-integration-dev-v2:test_big_query_dataset",
  "kmsKeyName": undefined,
  "location": "US",
  "name": "test_big_query_dataset",
  "public": false,
  "updatedOn": 1610919344642,
  "webLink": "https://console.cloud.google.com/bigquery?project=j1-gc-integration-dev-v2&d=test_big_query_dataset&p=j1-gc-integration-dev-v2&page=dataset",
}
`;

exports[`#createBigQueryDatasetEntity should convert to entity with public param set to on 1`] = `
{
  "_class": [
    "DataStore",
    "Database",
  ],
  "_key": "j1-gc-integration-dev-v2:test_big_query_dataset",
  "_rawData": [
    {
      "name": "default",
      "rawData": {
        "access": [
          {
            "role": "WRITER",
            "specialGroup": "allAuthenticatedUsers",
          },
          {
            "role": "WRITER",
            "specialGroup": "projectWriters",
          },
          {
            "role": "OWNER",
            "specialGroup": "projectOwners",
          },
          {
            "role": "OWNER",
            "userByEmail": "admin@creativice.com",
          },
          {
            "iamMember": "allUsers",
            "role": "READER",
          },
          {
            "role": "READER",
            "specialGroup": "projectReaders",
          },
        ],
        "creationTime": "1610919167542",
        "datasetReference": {
          "datasetId": "test_big_query_dataset",
          "projectId": "j1-gc-integration-dev-v2",
        },
        "etag": "Q1RJczpo1aoPRRoUJ1pElQ==",
        "id": "j1-gc-integration-dev-v2:test_big_query_dataset",
        "kind": "bigquery#dataset",
        "lastModifiedTime": "1610919344642",
        "location": "US",
        "selfLink": "https://www.googleapis.com/bigquery/v2/projects/j1-gc-integration-dev-v2/datasets/test_big_query_dataset",
      },
    },
  ],
  "_type": "google_bigquery_dataset",
  "classification": null,
  "createdOn": 1610919167542,
  "description": undefined,
  "displayName": "test_big_query_dataset",
  "encrypted": true,
  "etag": "Q1RJczpo1aoPRRoUJ1pElQ==",
  "id": "j1-gc-integration-dev-v2:test_big_query_dataset",
  "kmsKeyName": undefined,
  "location": "US",
  "name": "test_big_query_dataset",
  "public": true,
  "updatedOn": 1610919344642,
  "webLink": "https://console.cloud.google.com/bigquery?project=j1-gc-integration-dev-v2&d=test_big_query_dataset&p=j1-gc-integration-dev-v2&page=dataset",
}
`;

exports[`#createBigQueryModelEntity should convert to entity 1`] = `
{
  "_class": [
    "Model",
  ],
  "_key": "j1-gc-integration-dev-v2:test_big_query_dataset:natality_model",
  "_rawData": [
    {
      "name": "default",
      "rawData": {
        "creationTime": "1622485604349",
        "etag": "NEs92BeE7yNA56rXtlPojg==",
        "featureColumns": [
          {
            "name": "is_male",
            "type": {
              "typeKind": "BOOL",
            },
          },
          {
            "name": "gestation_weeks",
            "type": {
              "typeKind": "INT64",
            },
          },
          {
            "name": "mother_age",
            "type": {
              "typeKind": "INT64",
            },
          },
          {
            "name": "mother_race",
            "type": {
              "typeKind": "STRING",
            },
          },
        ],
        "labelColumns": [
          {
            "name": "predicted_weight_pounds",
            "type": {
              "typeKind": "FLOAT64",
            },
          },
        ],
        "lastModifiedTime": "1622485604453",
        "location": "US",
        "modelReference": {
          "datasetId": "natality",
          "modelId": "natality_model",
          "projectId": "j1-gc-integration-dev-v2",
        },
        "modelType": "LINEAR_REGRESSION",
        "trainingRuns": [
          {
            "dataSplitResult": {
              "evaluationTable": {
                "datasetId": "_dc91ff9c66b0b00709a1245f8bd9d5ddbfdf7aea",
                "projectId": "j1-gc-integration-dev-v2",
                "tableId": "anon74d97e3b_b105_4c61_a170_ab4d07a18969_imported_data_split_eval_data",
              },
              "trainingTable": {
                "datasetId": "_dc91ff9c66b0b00709a1245f8bd9d5ddbfdf7aea",
                "projectId": "j1-gc-integration-dev-v2",
                "tableId": "anon74d97e3b_b105_4c61_a170_ab4d07a18969_imported_data_split_training_data",
              },
            },
            "evaluationMetrics": {
              "regressionMetrics": {
                "meanAbsoluteError": 0.977932870859257,
                "meanSquaredError": 1.739664352683728,
                "meanSquaredLogError": 0.03609560638952963,
                "medianAbsoluteError": 0.7624851534167556,
                "rSquared": 0.043297645981164035,
              },
            },
            "results": [
              {
                "durationMs": "5022",
                "evalLoss": 1.7396643526837197,
                "index": 0,
                "trainingLoss": 1.6821061155036234,
              },
            ],
            "startTime": "2021-05-31T18:26:24.431Z",
            "trainingOptions": {
              "l2Regularization": 0,
              "lossType": "MEAN_SQUARED_LOSS",
              "optimizationStrategy": "NORMAL_EQUATION",
            },
          },
        ],
      },
    },
  ],
  "_type": "google_bigquery_model",
  "classification": null,
  "createdOn": 1622485604349,
  "description": undefined,
  "displayName": "natality_model",
  "etag": "NEs92BeE7yNA56rXtlPojg==",
  "expirationTime": undefined,
  "location": "US",
  "modelType": "LINEAR_REGRESSION",
  "name": "natality_model",
  "updatedOn": 1622485604453,
  "webLink": "https://console.cloud.google.com/bigquery?project=j1-gc-integration-dev-v2&d=natality&m=natality_model&p=j1-gc-integration-dev-v2&page=model",
}
`;

exports[`#createBigQueryTableEntity should convert to entity 1`] = `
{
  "_class": [
    "DataCollection",
  ],
  "_key": "j1-gc-integration-dev-v2:new_dataset_test.mascots",
  "_rawData": [
    {
      "name": "default",
      "rawData": {
        "creationTime": "1621610438123",
        "id": "j1-gc-integration-dev-v2:new_dataset_test.mascots",
        "kind": "bigquery#table",
        "tableReference": {
          "datasetId": "new_dataset_test",
          "projectId": "j1-gc-integration-dev-v2",
          "tableId": "mascots",
        },
        "type": "TABLE",
      },
    },
  ],
  "_type": "google_bigquery_table",
  "classification": null,
  "createdOn": 1621610438123,
  "displayName": "mascots",
  "expirationTime": undefined,
  "friendlyName": undefined,
  "id": "j1-gc-integration-dev-v2:new_dataset_test.mascots",
  "kmsKeyName": "example-kms-key-name",
  "name": "mascots",
  "public": false,
  "type": "TABLE",
  "webLink": "https://console.cloud.google.com/bigquery?project=j1-gc-integration-dev-v2&d=new_dataset_test&p=j1-gc-integration-dev-v2&t=mascots&page=table",
}
`;
